# Functions tuples and uniitests #5

### Functions

In [3]:
# function definition

def well_chosen_function_name(first_argument, second_argument):
    '''
    breif description of the function, its arguments and return value.
    some more details.
    '''
    local_varaible_name = 'Y' # 'visible' inside the function, but not outside
    return local_varaible_name * (first_argument + second_argument)

In [7]:
a = 2
b = 3

ab = well_chosen_function_name(a,b)
print(ab)

YYYYY


look at this. if you call the function, it returns local_variable_name, which here is 'Y', and * (first_agrument+second_argument) here is 5, so output would be 5 times of Y. YYYYY

### More on local variables

In [ ]:
x = 100
y = 5

def f(x):
    x -= y # x = x - y
    print("inside f:", x, y)

f(x)
print(x)
assert x == 100 # gives an error if it's not true; useful for ensuring correctness.

here, i assume the output would be "inside f: 95, 5". but was not able to guess 100.

review. 
1. before the function call, x = 100, y = 5
so in global scope: x = 100, y = 5.
2. call the function f(x)
so now global x = 100, local x = 100 (a copy of the value)
3. inside the function
equation modeifies the local x only, so x = 95 y = 5.
4. function ends, print("inside f: 95, 5)
5. print f(x) = 100

In [8]:
print(print("a"))

a
None


In [10]:
print(len("a"))

1


print() prints, it does not return.
when you nest it, you're printing None.

### Standard functions

* print
* help -- super useful
* abs, min, max -- basic numerical operations
* len, range... -- next time

In [12]:
help(well_chosen_function_name) # you can ask for help about any function (or class)

Help on function well_chosen_function_name in module __main__:

well_chosen_function_name(first_argument, second_argument)
    breif description of the function, its arguments and return value.
    some more details.



### Intermezzo: Unit tests
Testing is crucial in 'industrial' programming/software engineering
The simplest type of automated test is a unit test, in which you test a small unit (function/class) of your program.

pytest is a minimalistic (but poweful if needed) frameworkk of testing python code, see blow for an example

In [45]:
# we can place asserts in our code to make sure something is true
assert abs(-1) == 1

In [25]:
%%file to_test.py

# function to test
def double(x):
    '''
    returns twice the argument x; x is unchanged
    '''
    return x*2

# function to test
def test_double_on_zero():
    assert double(0) == 0
def test_double_on_integer():
    assert double(2) == 3
def test_double_on_string():
    assert double("a") == "aa"
def test_double_on_empty_string():
    assert double("") == ""    

Overwriting to_test.py


In [26]:
# function to test
def double(x):
    '''
    returns twice the argument x; x is unchanged
    '''
    return x*2

# function to test
def test_double_on_zero():
    assert double(0) == 0
def test_double_on_integer():
    assert double(2) == 3
def test_double_on_string():
    assert double("a") == "aa"
def test_double_on_empty_string():
    assert double("") == ""    

In [27]:
!cat to_test.py


# function to test
def double(x):
    '''
    returns twice the argument x; x is unchanged
    '''
    return x*2

# function to test
def test_double_on_zero():
    assert double(0) == 0
def test_double_on_integer():
    assert double(2) == 3
def test_double_on_string():
    assert double("a") == "aa"
def test_double_on_empty_string():
    assert double("") == ""    


In [28]:
!python -m pytest to_test.py

============================= test session starts ==============================
platform darwin -- Python 3.11.8, pytest-7.4.0, pluggy-1.0.0
rootdir: /Users/laurensong/Desktop/school/MAS4115
plugins: anyio-4.2.0
collected 4 items                                                              

to_test.py .F..                                                          [100%]

=================================== FAILURES ===================================
____________________________ test_double_on_integer ____________________________

    def test_double_on_integer():
>       assert double(2) == 3
E       assert 4 == 3
E        +  where 4 = double(2)

to_test.py:13: AssertionError
=========================== short test summary info ============================
FAILED to_test.py::test_double_on_integer - assert 4 == 3
========================= 1 failed, 3 passed in 0.04s ==========================


In [71]:
%%file to_test_triple.py

# function to test
def triple(x):
    '''
    returns twice the argument x; x is unchanged
    '''
    return x*3

# function to test
def test_cube_on_zero():
    assert triple(0) == 0
def test_cube_on_integer():
    assert triple(2) == 6
def test_cube_on_string():
    assert triple("a") == "aaa"
def test_cube_on_empty_string():
    assert triple("") == ""    

Writing to_test_triple.py


In [72]:
# function to test
def triple(x):
    '''
    returns twice the argument x; x is unchanged
    '''
    return x*3

# function to test
def test_cube_on_zero():
    assert triple(0) == 0
def test_cube_on_integer():
    assert triple(2) == 6
def test_cube_on_string():
    assert triple("a") == "aaa"
def test_cube_on_empty_string():
    assert triple("") == ""  

In [73]:
!cat to_test_triple.py


# function to test
def triple(x):
    '''
    returns twice the argument x; x is unchanged
    '''
    return x*3

# function to test
def test_cube_on_zero():
    assert triple(0) == 0
def test_cube_on_integer():
    assert triple(2) == 6
def test_cube_on_string():
    assert triple("a") == "aaa"
def test_cube_on_empty_string():
    assert triple("") == ""    


In [74]:
!python -m pytest to_test_triple.py

============================= test session starts ==============================
platform darwin -- Python 3.11.8, pytest-7.4.0, pluggy-1.0.0
rootdir: /Users/laurensong/Desktop/school/MAS4115
plugins: anyio-4.2.0
collected 4 items                                                              

to_test_triple.py ....                                                   [100%]

============================== 4 passed in 0.01s ===============================


so, unit test actually is: answer the simple question "given this input, my code produce the correct output?"

unit = a small piece of code (usually a function)
test = an automatic ccheck using **assert**

In [21]:
assert 2 + 2 == 5

AssertionError: 

like code above, assert means "condition." so, if condition is true, nothing happens other wise program crashes loudly.

why pytest recognizes this automatically? 

pytest looks for: 
* file name test_*.py or *_test.py
* functions named test_*

### Tuples
in python tuples are used to store a (small) bunch of values of possbile different type. very useful for returning more than one element from a function.


In [75]:
def give_three_numbers():
    return 1, 2, 3.14 # this creates and returns a tuple

In [83]:
t = give_three_numbers() # t is tuple
print(type(t)) # type of t = tuple
print(t)
print(t[0] + t[1] + t[2]) # we can index elements of the tuple

a, b, c = give_three_numbers() # automatic unpacking of a tuple!
print(a, b, c) 

first, *rest = give_three_numbers() # slightly more advanced

<class 'tuple'>
(1, 2, 3.14)
6.140000000000001
1 2 3.14


here, you might wonder why print(t[0] + t[1] + t[2]) does not print 6.14 but print 6.140000000000001.
computer store floating-point numbers in binary, not decimal.

* decimal fractions -> base-10
* binary fractions -> base-2
some decimals are impossible to represent exactly in base-2

so python stores the closest representable binary number, not the exact decimal value.

also, for the difference between 
**a, b, c = give_three_numbers()** and **first, *rest = give_three_numbers()** is 
first one should care about the count, otherwise it gets value error
however, the second one is care only one variable may use * to capture.
all remaining elements into a list. so, counts don't have to match exactly.

In [84]:
a, b, c = t # unpacking from a variable
print(a, b, c)

1 2 3.14


In [87]:
def print_three(a, b, c):
    print('Here is your triple:', a, b, c)

t = 1, 2, 3
print_three(*t) # we can push a tuple as argumnets a, b, c

Here is your triple: 1 2 3


In [91]:
# then you might want to do with your birthday!

def print_your_birthday(month, date, year):
    print("Your birthday is:", month, date, year)

b = 2, 19, 2003
print(*b)


2 19 2003


### Task 0

Write a function which takes to vectors a, b in R^2 and returns a vevtor t such that if we add t to a, we will ge b. 

We can represent each vector as a tuple (of size 2)

Write a coiple of unit tests and run them!

first, i tried with abs(b-a) however, then a and b cannot be vector(tuple). a b are scalars.
so b - a is unavailable in tuple.

correct way is using tuple! + index

In [146]:
%%file x.py

def sub(a,b):
    '''
    Remember that both a and b are vectors, represented as tuples (a-b)
    '''
    return (b[0] - a[0], b[1] - a[1])


def test_sub_zero():
    zero_vector = (0,0)
    assert sub(zero_vector, zero_vector) == zero_vector

def test_sub_positive():
    a = (4, 2)
    b = (6, 4)
    assert sub(a,b) == (2,2)

def test_sub_negative():
    a = (-1, 1)
    b = (-2, 3)
    assert sub(a,b) == (-1,2)




Overwriting x.py


In [147]:
!cat x.py


def sub(a,b):
    '''
    Remember that both a and b are vectors, represented as tuples (a-b)
    '''
    return (b[0] - a[0], b[1] - a[1])


def test_sub_zero():
    zero_vector = (0,0)
    assert sub(zero_vector, zero_vector) == zero_vector

def test_sub_positive():
    a = (4, 2)
    b = (6, 4)
    assert sub(a,b) == (2,2)

def test_sub_negative():
    a = (-1, 1)
    b = (-2, 3)
    assert sub(a,b) == (-1,2)




In [148]:
!python -m pytest x.py

============================= test session starts ==============================
platform darwin -- Python 3.11.8, pytest-7.4.0, pluggy-1.0.0
rootdir: /Users/laurensong/Desktop/school/MAS4115
plugins: anyio-4.2.0
collected 3 items                                                              

x.py ...                                                                 [100%]

============================== 3 passed in 0.01s ===============================


### Default arguments
it's often convenient to have default values to some of the arguments
if we want to override an default argument value, it's often clearer if we explicitly spell out the name.

In [149]:
# 1e-8 means 10**-8, see https://en.wikipedia.org/wiki/Scientific_notation
def are_epsilon_close(a, b, eps=1e-8):
  return abs(a-b) <= eps

assert are_epsilon_close((2**0.5)**2, 2) # using the default eps
assert are_epsilon_close((2**0.5)**2, 2, eps=1e-12) # overriding eps
# assert are_epsilon_close((2**0.5)**2, 2, eps = 0) # fails!

### Functins taking functions as arguments (intermediate)

it's often useful to pass one function to another.
for example, min/mas function accept an optional parameter 'key', which is a function which should extract the value to compare the elements by.